nixos/modules/hardware/iosched.nix

```diff
@@ -121,19 +137,33 @@ in
           (pkgs.writeTextDir "etc/udev/rules.d/98-block-io-scheduler.rules" (
             concatLines (
               map (concatStringsSep ", ") (
-                optional (cfg.defaultScheduler != null) [
-                  ''SUBSYSTEM=="block"''
-                  ''ACTION=="add|change"''
-                  ''TEST=="queue/scheduler"''
-                  ''ATTR{queue/scheduler}="${escape cfg.defaultScheduler}"''
-                ]
-                ++ optional (cfg.defaultSchedulerRotational != null) [
-                  ''SUBSYSTEM=="block"''
-                  ''ACTION=="add|change"''
-                  ''ATTR{queue/rotational}=="1"''
-                  ''TEST=="queue/scheduler"''
-                  ''ATTR{queue/scheduler}="${escape cfg.defaultSchedulerRotational}"''
-                ]
+                optional (cfg.defaultScheduler != null) (
+                  [
+                    ''SUBSYSTEM=="block"''
+                    ''ACTION=="add|change"''
+                  ]
+                  ++ optional (
+                    cfg.defaultSchedulerExclude != null
+                  ) ''KERNEL!="${escape cfg.defaultSchedulerExclude}"''
+                  ++ [
+                    ''TEST=="queue/scheduler"''
+                    ''ATTR{queue/scheduler}="${escape cfg.defaultScheduler}"''
+                  ]
+                )
```

Comment:
I found this code very hard to follow with the nested optionals.

I also noticed that it's repeated; the first three elements are the same. Perhaps they could be extracted into a constant and then simply applied something like this:

```Nix
header ++ optionals (cfg...) [ ... ]
```

Reply:
I moved the rule construction into a `udevRule` function to make the code easier to understand.

Reply:
Much better, thank you :)